### PR TITLE
fix(tools): tell the model an invented tool does not exist, not that it lacks permission

### PR DESCRIPTION
**Pull request by griffinwork40**
## Summary

Subagents kept emitting `str_replace_based_edit_tool` and getting back:

```
Tool "str_replace_based_edit_tool" is not in the configured allowlist
```

That name is not a hallucination. It is a **real Anthropic tool name** — the text editor tool, types `text_editor_20250429` / `text_editor_20250728`, Claude 4 and later. This project does not implement it (it ships `edit_file`), but the models carry a strong post-training prior for the native name and re-emit it during long, edit-heavy runs.

The message they got back is the problem. `SessionToolDispatcher` has two rejection paths:

| Gate | Message |
|---|---|
| Permission check (`runPreDispatchGates`) | `Tool "X" is not in the configured allowlist` |
| Handler lookup (`executeCoreInner`) | `Unknown tool "X". Available tools: ...` |

The permission gate runs **first** and returns early. Since every surface configures an allowlist (`CHILD_ALLOWED_TOOLS`, `topLevelSurfaceAllowedTools`), a name the model invented always short-circuits there — so the branch that would have told it the tool does not exist, and named the real ones, was **structurally unreachable for exactly the case it was written for**. It only ever fired for a name that *is* allowlisted but has no handler.

So the model saw what looks like a permissions problem, had no signal the tool was fictional, and no list of what to use instead. Each attempt burns a round against `max_tool_use_iterations` and feeds the repeat-failure guard.

This is the same failure shape already documented in `top-level-allowlist.ts` — where `get_runtime_state` and `exit_plan_mode` were rejected as "not in the configured allowlist" while visibly callable. That fix centralized the list; the message masking itself was never addressed.

### Commit 1 — discriminate "does not exist" from "not permitted"

Handler presence is the discriminator. A **registered-but-denied** tool is a real permission decision and keeps its original message and `failureClass` untouched. An **unregistered** name now gets the callable-tool list plus an explicit do-not-retry instruction.

`failureClass` deliberately stays `permission-denied`: it is a trace union consumed by receipt/detector code, and widening it is out of scope for a message fix.

### Commit 2 — stop the unknown-tool message leaking non-allowlisted handlers

The handler-lookup branch enumerated raw `this.handlers.keys()`, which can contain tools the gate will reject — suggesting one just buys a denial next turn. This path is reachable in production, not theoretical: `exit_plan_mode` is listed statically by `topLevelSurfaceAllowedTools` but only registered while in plan mode, and an MCP tool whose server dropped after the allowlist snapshot hits the same branch.

Both sites now share one `unknownToolMessage()` helper sourced from `toolDefs`, honouring the contract already written on that getter — *never show the model a tool the permission gate will reject* — so the two can no longer drift.

### Why not just implement the tool

Considered and rejected:

- **Breaks the provider-agnostic contract.** `text_editor_20250728` is an Anthropic-typed tool with no OpenAI equivalent. There is no precedent — no native tool type is passed anywhere today. The child tool surface would silently differ by provider.
- **It is not one duplicate, it is three.** The tool dispatches on a `command` field spanning `view` → `read_file`, `create` → `write_file`, `str_replace`/`insert` → `edit_file`, making tool selection *more* nondeterministic.
- **Anthropic-versioned schema.** The name already changed once (`str_replace_editor` → `str_replace_based_edit_tool`) and `undo_edit` was dropped in `20250429`.
- **The pain does not justify it.** Local telemetry shows ~44 occurrences total, self-correcting to `edit_file` within 1–3 calls essentially every time.

The message fix also generalizes: it repairs the whole class, so any mistyped or invented tool name self-corrects — including whatever a future model generation invents.

## How was this tested?

- `pnpm lint` — clean.
- `pnpm test` — **833/833 files, 15864 passed, 2 skipped.** Baseline at `8ba6e2cd` is 15857 passed; the delta is exactly the 7 net-new tests here.
- `pnpm build`, `audit:sdk:check`, `audit:env:check`, `scan:env:check`, `audit:chalk:check`, `fix:pins:check` — all pass.
- 7 net-new tests. This path previously had **no coverage** for allowlist denial; the pre-existing "returns isError for unknown tool" test covers the other branch and still passes unchanged.

Tests cover: the phantom name no longer blamed on the allowlist; the callable-tool list and do-not-retry guidance; a registered-but-denied tool keeping its original message (the safety property); `failureClass` preserved; the non-leak guarantee on the allowlisted-but-unregistered path; and the empty-listing edge.

One full-suite run showed a failure in `subagent-output-capture-wiring.test.ts` — verified pre-existing and unrelated: it passes in isolation, a stashed pristine-`HEAD` baseline passed 833/833, and re-running with this diff restored passed 833/833.

## Checklist

- [x] `pnpm lint` passes
- [x] `pnpm test` passes
- [x] Commits are signed off (`git commit -s`) per the [DCO](https://developercertificate.org/) — see [CONTRIBUTING.md](../CONTRIBUTING.md)
- [x] No secrets, tokens, or private paths in the diff

### CODE REVIEW

File: src/agent/tools/dispatcher.ts [hunk 1]
```diff
@@ -554,6 +554,65 @@ export class SessionToolDispatcher implements ToolDispatcher {
...
+   *
+   * `failureClass` stays `permission-denied` at the call site: it is a trace
+   * union consumed by receipt/detector code, and widening it is out of scope
+   * for a message fix.
+   */
+  private denialReason(toolName: string, permissionReason: string | undefined): string {
+    if (this.handlers.has(toolName)) {
+      return permissionReason ?? `Tool "${toolName}" is not permitted`;
+    }
+    return this.unknownToolMessage(toolName);
```
[the remark this answers is not in the document]
Comment (griffinwork40) [reply src/agent/tools/dispatcher.ts]: @codex fix comments

**Comment by griffinwork40** [comment]
**Wave 1.5 complete.** Results:

**Check A (citations):** All cited line ranges verified against ref `c61d336140215ad18cf78a42704ef787b9026394`. All match.

**Check B (absence claims):**
- F4 claimed "`withMcpToolsAllowed` does not exist in `permissions.ts`" → **false-absent**: function exists at `src/agent/tools/permissions.ts:60` (`export function withMcpToolsAllowed(`). **F4 DROPPED.**

**Verification manifest:**
```
[{type: citation, claim: "emitPreToolUseBlock at line 944-947", status: verified, finding_id: F1}]
[{type: citation, claim: "handler-not-found return at 1316-1317", status: verified, finding_id: F2}]
[{type: citation, claim: "emitPreToolUseBlock absent at 1314-1318", status: verified, finding_id: F3}]
[{type: absence, claim: "withMcpToolsAllowed absent from permissions.ts", status: false-absent, evidence: "permissions.ts:60: export function withMcpToolsAllowed(", finding_id: F4, action: DROPPED}]
[{type: citation, claim: "toolDefs getter at 550-555", status: verified, finding_id: F5}]
```

No critical/high findings and no overridden medium findings → shadow-verify not triggered. Proceeding to synthesis.

---

## PR #983 Review — `fix(tools): tell the model an invented tool does not exist, not that it lacks permission`

**Reviewed ref:** `c61d336140215ad18cf78a42704ef787b9026394` · Branch `afk/debug-subagent-tool-calls → main`  
**Regime:** light (199 lines total, hotfix)  
**Stated intent:** PR title + body — spec-compliance assessed.

---

### Findings

---

**F2 · medium · blocking: true · high confidence · correctness / observability**  
`src/agent/tools/dispatcher.ts:1316–1317` · ref: c61d336 · `diff-context`

The handler-not-found return path carries no `failureClass`:

```typescript
return { content: this.unknownToolMessage(call.name), isError: true };
```

`BENIGN_FAILURE_CLASSES` (trace/types.ts:132–134) includes `permission-denied` and excludes unclassified results. Per the JSDoc at trace/types.ts:129: *"An unclassified failure (no `failureClass`) is never benign."* The `exit_plan_mode`-in-wrong-mode and MCP-server-drop cases that the PR docstring explicitly documents as production-reachable via this path produce `isError: true` with no class, landing in the alarming (non-benign) bucket of `tool-failure-density` telemetry. The PR unifies both call sites under `unknownToolMessage` and documents this path as intentional, making the gap newly salient even though it is pre-existing on this branch.

**Suggestion:** Add `failureClass: 'permission-denied' as const` to the handler-not-found return at line 1317 — a two-word fix that belongs in this PR rather than a follow-up, given the PR already names and documents this path. Alternatively introduce `'missing-handler'` in `TOOL_FAILURE_CLASSES` and add it to `BENIGN_FAILURE_CLASSES` if the distinction matters to detectors.

---

**F3 · low · blocking: false · high confidence · correctness**  
`src/agent/tools/dispatcher.ts:1316–1317` · ref: c61d336 · `diff-context`

The handler-not-found return does not call `emitPreToolUseBlock`. The invariant at dispatcher.ts:670–678 states: *"EVERY path that returns an isError ToolResult to the model for a POLICY reason must call this."* An allowlisted name with no registered handler (exit_plan_mode, MCP-server-drop) is a policy-adjacent denial. The trace has no durable record this denial occurred. Pre-existing gap, but the PR documents both paths and would be the natural place to close it.

**Suggestion:** Add `await this.emitPreToolUseBlock(call.name, this.unknownToolMessage(call.name))` before the return at line 1317, matching the pattern used at the permission gate (lines 945–948) and the read-only bash gate.

---

**F4 · DROPPED** — `withMcpToolsAllowed` absence claim refuted: function exists at `src/agent/tools/permissions.ts:60`. JSDoc citation is accurate, not dangling.

---

**F1 · nit · blocking: false · high confidence · correctness (verification)**  
`src/agent/tools/dispatcher.ts:944–948` · ref: c61d336 · `file-state`

The diff elides `emitPreToolUseBlock` from the permission-check hunk, creating a plausible reader concern that the trace-emission call was dropped. It was not — the call fires at line 946 with the new `reason` string and the trace invariant is upheld. No action needed.

---

**F5 · nit · blocking: false · high confidence · correctness**  
`src/agent/tools/dispatcher.ts:550–555` · ref: c61d336 · `file-state`

`toolDefs` correctly returns only schemas whose names are in the allowlist. The old handler-lookup error advertised raw `this.handlers.keys()` (including non-allowlisted tools). The replacement `this.toolDefs` is the right source — matches exactly what the model was shown. Clean.

---

**F6 · nit · blocking: false · high confidence · test-coverage**  
`src/agent/tools/dispatcher.test.ts:177–196` · ref: c61d336 · `diff-context`

The `it.each` test for executor-backed tools (`agent`/`skill`/`compose`) correctly exercises `isRegisteredTool` via the executor option detection and the permission gate. The `not.toContain('Unknown tool')` assertion correctly validates the branch split. Valid test.

---

**F8 · nit · blocking: false · high confidence · test-coverage**  
`src/agent/tools/dispatcher.test.ts:219–233` · ref: c61d336 · `diff-context`

The empty-`toolDefs` guard (omitting "listed above" / "Available tools:") is tested only via the permission-gate path (`ghost` test). The handler-not-found path with an empty `toolDefs` is untested. Vanishingly narrow edge case (requires allowlist-to-toolDefs mismatch), not blocking.

---

### Spec-compliance

**Assessed.** Stated intent: tell the model an invented tool does not exist, not that it lacks permission.

- ✅ Unregistered names at the permission gate now receive `unknownToolMessage` with the callable-tool list and do-not-retry guidance.
- ✅ Registered-but-denied tools preserve their original message and `failureClass`.
- ✅ Both call sites (permission gate + handler lookup) are unified under `unknownToolMessage`.
- ✅ `toolDefs` used (not `handlers.keys()`) so denied tools are never suggested.
- No scope creep observed.

---

### Security / api-compat

No issues. The change narrows message surface (less information given to models about the internal handler map). No auth surface, no secret exposure, no breaking API change.

---

### Perf / observability

No hot-path impact. The three new private methods (`denialReason`, `isRegisteredTool`, `unknownToolMessage`) run only on denial paths, not on the successful tool dispatch path. The telemetry gap noted in F2 is the only observability concern.

---

### What was not checked

- Citations verified inline against branch HEAD `c61d336140215ad18cf78a42704ef787b9026394`.
- Stated intent: PR title + body — spec-compliance assessed.
- Did not run the test suite (read-only review).
- Did not review MCP server integration layer beyond checking `isRegisteredTool` coverage.
- Did not review Telegram surface, daemon, or REPL invocation paths.

---

### Decision: DO NOT MERGE — 1 medium blocking (correctness); 2 low, 3 nit non-blocking.

**F2** (no `failureClass` on handler-not-found path → unclassified failures in telemetry) is a two-word fix the PR itself creates the opportunity for by documenting this path as production-reachable. The message fix is correct and complete. Resolving F2 before merge is the only gate.

**Done**
- Read PR #983 diff, fetched stated intent and HEAD SHA `c61d336`
- Dispatched Wave 1 research-agent (light regime, all dimensions)
- Ran Wave 1.5 inline: verified all file-state citations at the reviewed ref; caught one false-absent claim (F4 dropped — `withMcpToolsAllowed` exists at `permissions.ts:60`)
- No critical/high findings → shadow-verify not triggered per protocol
- Synthesized final review report above
- Files changed in this session: none (read-only review)

---
_🤖 Posted by `agent-afk /review --post github`_
